File: examples/api-consumer.py
#!/usr/bin/env python3
"""
Python API Consumer for Microservices
---------------------------------------------------------------------
Este script actúa como un cliente que consume dos APIs distintas:
 - Una API principal desarrollada en Laravel.
 - Una API de procesamiento intermedio desarrollada en Node.js.

El objetivo es demostrar cómo un microservicio puede comunicarse con
otros servicios, obtener datos, procesarlos y crear nuevos registros.
---------------------------------------------------------------------
"""

import requests
import json


class ProductAPIConsumer:
    def __init__(self):
        """
        Constructor básico. Configura las URLs base de las APIs
        que serán consumidas por el cliente Python.
        """
        self.laravel_api = "http://localhost:8000/api"
        self.node_api = "http://localhost:3001/api"
        print("Product API Consumer initialized.")


    def test_health(self):
        """
        Comprueba el estado del microservicio en Node.js.
        Retorna un diccionario con la respuesta o None si falla.
        """
        try:
            response = requests.get(f"{self.node_api}/health")
            data = response.json()
            print(f"Health Check: {data['message']}")
            return data
        except Exception as e:
            print(f"Health Check Failed: {e}")
            return None


    def get_products_from_laravel(self):
        """
        Obtiene productos sin procesar directamente desde Laravel.
        Retorna una lista con los productos o una lista vacía si ocurre un error.
        """
        try:
            response = requests.get(f"{self.laravel_api}/products")
            data = response.json()
            print(f"Laravel Products: {len(data['data'])} found")
            return data["data"]
        except Exception as e:
            print(f"Laravel API Error: {e}")
            return []


    def get_processed_products(self):
        """
        Obtiene productos procesados por el microservicio Node.js.
        La data incluye resumen, países involucrados y normalización extra.
        """
        try:
            response = requests.get(f"{self.node_api}/products")
            data = response.json()
            print(f"Node.js Processed Products: {data['summary']['total_count']} total")
            print(f"Countries involved: {data['summary']['countries']}")
            return data
        except Exception as e:
            print(f"Node.js API Error: {e}")
            return None


    def create_product(self, name, country_code):
        """
        Envía un POST de creación de producto al microservicio Node.js.
        Retorna el producto creado o None si ocurre un error.
        """
        try:
            payload = {
                "name": name,
                "country_code": country_code
            }

            response = requests.post(
                f"{self.node_api}/products",
                json=payload,
                headers={"Content-Type": "application/json"}
            )

            data = response.json()

            if data.get("success"):
                print(f"Product Created: {data['data']['name']}")
                print(f"SKU: {data['data']['sku']}")
                print(f"Country: {data['data']['country']['name']}")
                return data["data"]
            else:
                print(f"Creation Failed: {data.get('message')}")
                return None

        except Exception as e:
            print(f"API Error: {e}")
            return None


    def run_demo(self):
        """
        Ejecuta una demostración completa del flujo:
         - Verificar salud del servicio Node.js
         - Obtener datos en bruto de Laravel
         - Obtener datos procesados desde Node.js
         - Crear un producto de prueba
         - Mostrar un producto de ejemplo desde Laravel
        """
        print("\nStarting Microservices Demo\n")

        # Health check
        self.test_health()

        # Get raw data
        laravel_products = self.get_products_from_laravel()

        # Processed data
        self.get_processed_products()

        # Create a new product
        new_product = self.create_product("Python Demo Product", "BR")

        # Display sample product
        if laravel_products:
            print("\nSample Laravel Product:")
            sample = laravel_products[0]
            print(f"Name: {sample['name']}")
            print(f"SKU: {sample['sku']}")
            print(f"Country: {sample['country_code']}")

        print("\nDemo complete. Microservices are communicating correctly.")


if __name__ == "__main__":
    consumer = ProductAPIConsumer()
    consumer.run_demo()

File: examples/api-consumer.js
// API Consumer Example - JavaScript

const LARAVEL_API = 'http://localhost:8000/api';
const NODE_API = 'http://localhost:3001/api';

/**
 * Clase encargada de consumir la API principal de Laravel
 * y el microservicio Node.js usando fetch.
 *
 * Este cliente permite:
 *  - Probar el estado de los servicios
 *  - Obtener productos desde Laravel
 *  - Obtener productos procesados desde Node.js
 *  - Crear productos mediante Node.js
 *  - Filtrar productos por país
 *  - Obtener estadísticas combinadas
 */
class ProductAPIConsumer {
    constructor() {
        console.log('Product API Consumer Initialized');
    }

    /**
     * Verifica la salud del microservicio Node.js.
     */
    async testHealth() {
        try {
            const response = await fetch(`${NODE_API}/health`);
            const data = await response.json();
            console.log('Health Check:', data.message);
            return data;
        } catch (error) {
            console.error('Error en Health Check:', error.message);
        }
    }

    /**
     * Obtiene la lista de productos directamente desde la API de Laravel.
     */
    async getProductsFromLaravel() {
        try {
            const response = await fetch(`${LARAVEL_API}/products`);
            const data = await response.json();
            console.log(`Laravel Products: ${data.data.length} productos`);
            return data.data;
        } catch (error) {
            console.error('Error al consumir Laravel API:', error.message);
        }
    }

    /**
     * Obtiene productos procesados y normalizados por el microservicio Node.js.
     */
    async getProcessedProductsFromNode() {
        try {
            const response = await fetch(`${NODE_API}/products`);
            const data = await response.json();
            console.log(`Node.js Processed Products: ${data.summary.total_count} productos`);
            console.log('Países detectados:', data.summary.countries);
            return data;
        } catch (error) {
            console.error('Error al consumir Node.js API:', error.message);
        }
    }

    /**
     * Crea un nuevo producto por medio del microservicio Node.js.
     *
     * @param {string} name - Nombre del producto
     * @param {string} countryCode - Código de país (ISO)
     */
    async createProduct(name, countryCode) {
        try {
            const response = await fetch(`${NODE_API}/products`, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({ name, country_code: countryCode })
            });

            const data = await response.json();
            if (data.success) {
                console.log(`Producto creado: ${data.data.name}`);
                console.log(`SKU: ${data.data.sku}`);
                console.log(`País: ${data.data.country.name}`);
                return data.data;
            } else {
                console.error('Error al crear producto:', data.message);
            }
        } catch (error) {
            console.error('Error en API:', error.message);
        }
    }

    /**
     * Obtiene productos filtrados por país utilizando el microservicio Node.js.
     *
     * @param {string} countryCode - Código de país (ej. "US", "CA")
     */
    async getProductsByCountry(countryCode) {
        try {
            const response = await fetch(`${NODE_API}/country/${countryCode}/products`);
            const data = await response.json();
            console.log(`Productos de ${countryCode}: ${data.country.total_products} productos`);
            return data.data;
        } catch (error) {
            console.error('Error al filtrar productos por país:', error.message);
        }
    }

    /**
     * Obtiene estadísticas combinadas generadas por el microservicio Node.js.
     */
    async getStatistics() {
        try {
            const response = await fetch(`${NODE_API}/products/stats`);
            const data = await response.json();
            console.log('Estadísticas:');
            console.log('Total de productos:', data.data.total_products);
            console.log('Por país:', JSON.stringify(data.data.by_country, null, 2));
            return data.data;
        } catch (error) {
            console.error('Error al obtener estadísticas:', error.message);
        }
    }

    /**
     * Ejecuta una demostración completa de todas las funciones disponibles.
     */
    async runFullDemo() {
        console.log('Iniciando Demo de Microservicios\n');

        await this.testHealth();
        await this.getProductsFromLaravel();
        await this.getProcessedProductsFromNode();
        await this.createProduct('Demo Product', 'CA');
        await this.getProductsByCountry('US');
        await this.getStatistics();

        console.log('\nDemo completada.');
        console.log('Los microservicios se están comunicando correctamente.');
    }
}

// Ejecuta la demo si este archivo se corre directamente desde Node.js
if (typeof require !== 'undefined' && require.main === module) {
    const consumer = new ProductAPIConsumer();
    consumer.runFullDemo().catch(console.error);
}

module.exports = ProductAPIConsumer;
